internal/query/relations.go

```diff
@@ -267,16 +269,166 @@ func (e *engine) CreateRelationships(ctx context.Context, rels []types.Relations
-	role := newRole(actions)
+func (e *engine) CreateRole(ctx context.Context, actor, res types.Resource, roleName string, actions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.CreateRole")
+
+	defer span.End()
+
+	role := newRole(roleName, actions)
 	roleRels := e.roleRelationships(role, res)
 
+	dbTx, err := e.db.CreateRoleTransaction(ctx, actor.ID, role.ID, roleName, res.ID)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	defer dbTx.Rollback() //nolint:errcheck // error is logged in function
+
 	request := &pb.WriteRelationshipsRequest{Updates: roleRels}
 
 	if _, err := e.client.WriteRelationships(ctx, request); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.Role{}, err
+	}
+
+	if err = dbTx.Commit(); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
 		return types.Role{}, err
 	}
 
+	role.Creator = dbTx.Record.CreatorID
+	role.ResourceID = dbTx.Record.ResourceID
+	role.CreatedAt = dbTx.Record.CreatedAt
+	role.UpdatedAt = dbTx.Record.UpdatedAt
+
+	return role, nil
+}
+
+// actionsDiff determines which actions needs to be added and removed.
+// If no new actions are provided it is assumed no changes are requested.
+func actionsDiff(oldActions, newActions []string) ([]string, []string) {
+	if len(newActions) == 0 {
+		return nil, nil
+	}
+
+	old := make(map[string]bool, len(oldActions))
+	new := make(map[string]bool, len(newActions))
+
+	var add, rem []string
+
+	for _, action := range oldActions {
+		old[action] = true
+	}
+
+	for _, action := range newActions {
+		new[action] = true
+
+		// If the new action is not in the old actions, then we need to add the action.
+		if !old[action] {
+			add = append(add, action)
+		}
+	}
+
+	for _, action := range oldActions {
+		// If the old action is not in the new actions, then we need to remove it.
+		if !new[action] {
+			rem = append(rem, action)
+		}
+	}
+
+	return add, rem
+}
+
+// UpdateRole allows for updating an existing role with a new name and new actions.
+// If new name is empty, no change is made.
+// If new actions is an empty slice, no change is made.
+func (e *engine) UpdateRole(ctx context.Context, actor, roleResource types.Resource, newName string, newActions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.UpdateRole")
+
+	defer span.End()
+
+	role, err := e.GetRole(ctx, roleResource)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	newName = strings.TrimSpace(newName)
+
+	addActions, remActions := actionsDiff(role.Actions, newActions)
+
+	// If no changes, return existing role with no changes.
+	if role.Name == newName && len(addActions) == 0 && len(remActions) == 0 {
+		return role, nil
+	}
+
+	resourceID := role.ResourceID
+
+	// If the resource id is not found in the permissions database, then we must locate it in the spicedb database.
```

Comment:
If this is the case, don't we have an error in our logic? No role should exist without a corresponding row in the main CRDB database.

Reply:
Unless we provide a migration process that loads all existing roles from spicedb and populates crdb. Initially crdb will be empty. The expectation was that we would allow updating of an existing spicedb role with a name. The role has to exist to pass permission checks.